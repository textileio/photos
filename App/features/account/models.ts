import { ICafeSession } from '@textile/react-native-sdk'

export interface CafeSessionData {
  readonly session?: ICafeSession
  readonly processing: boolean
  readonly error?: string
}

export interface CafeSessionsData {
  readonly [peerId: string]: CafeSessionData | undefined
}
