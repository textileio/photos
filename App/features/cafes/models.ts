import { ICafeSession } from '@textile/react-native-sdk'

export interface Cafe {
  readonly peerId: string
  readonly token?: string // this is optional because we may not have tokens for previously registered custom cafes
  readonly error?: string
  readonly state: 'registering' | 'registered' | 'deregistering'
}

export interface Cafes {
  readonly [peerId: string]: Cafe
}

export interface CafeSessionData {
  readonly session: ICafeSession
  readonly processing: boolean
  readonly error?: string
}

export interface CafeSessionsData {
  readonly [peerId: string]: CafeSessionData
}
