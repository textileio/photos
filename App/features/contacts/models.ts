import { pb } from '@textile/react-native-sdk'
import Contacts from 'react-native-contacts'

export interface AddingContact {
  readonly error?: string
}

export interface AddingContacts {
  readonly [key: string]: AddingContact
}

export interface TextileSearchResult {
  readonly key: string
  readonly type: 'textile'
  readonly data: {
    contact: pb.IContact,
    isContact: boolean,
    adding: boolean
  }
}

export interface AddressBookSearchResult {
  readonly key: string
  readonly type: 'addressBook'
  readonly data: Contacts.Contact
}

export interface ErrorSearchResult {
  readonly key: string
  readonly type: 'error'
  readonly data: string
}

export interface LoadingSearchResult {
  readonly key: string
  readonly type: 'loading'
}

export interface EmptySearchResult {
  readonly key: string
  readonly type: 'empty'
}

export type SearchResult = TextileSearchResult | AddressBookSearchResult | ErrorSearchResult | LoadingSearchResult | EmptySearchResult

export interface SearchResultsSection {
  readonly key: string
  readonly title: string
  readonly data: SearchResult[]
}
