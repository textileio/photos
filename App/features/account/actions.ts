import { createAction, createAsyncAction } from 'typesafe-actions'
import { IContact } from '@textile/react-native-sdk'

import { SharedImage } from '../group/add-photo/models'

export const chooseProfilePhoto = createAsyncAction(
  'account/CHOOSE_PROFILE_PHOTO_REQUEST',
  'account/CHOOSE_PROFILE_PHOTO_SUCCESS',
  'account/CHOOSE_PROFILE_PHOTO_ERROR'
)<void, { image: SharedImage; data: string }, any>()

export const cancelProfilePhotoUpdate = createAction(
  'account/CANCEL_PROFILE_PHOTO_UPDATE',
  resolve => {
    return () => resolve()
  }
)

export const refreshProfileRequest = createAction(
  'account/REFRESH_PROFILE_REQUEST'
)

export const refreshProfileSuccess = createAction(
  'account/REFRESH_PROFILE_SUCCESS',
  resolve => {
    return (profile: IContact) => resolve({ profile })
  }
)

export const profileError = createAction('account/PROFILE_ERROR', resolve => {
  return (error: any) => resolve({ error })
})

export const refreshPeerIdRequest = createAction(
  'account/REFRESH_PEER_ID_REQUEST'
)

export const refreshPeerIdSuccess = createAction(
  'account/REFRESH_PEER_ID_SUCCESS',
  resolve => {
    return (peerId: string) => resolve({ peerId })
  }
)

export const refreshPeerIdError = createAction(
  'account/REFRESH_PEER_ID_ERROR',
  resolve => {
    return (error: any) => resolve({ error })
  }
)

export const refreshAddressRequest = createAction(
  'account/REFRESH_ADDRESS_REQUEST'
)

export const refreshAddressSuccess = createAction(
  'account/REFRESH_ADDRESS_SUCCESS',
  resolve => {
    return (address: string) => resolve({ address })
  }
)

export const refreshAddressError = createAction(
  'account/REFRESH_ADDRESS_ERROR',
  resolve => {
    return (error: any) => resolve({ error })
  }
)

export const setUsernameRequest = createAction(
  'account/SET_USERNAME_REQUEST',
  resolve => {
    return (username: string) => resolve({ username })
  }
)

export const setAvatar = createAsyncAction(
  'account/SET_AVATAR_REQUEST',
  'account/SET_AVATAR_SUCCESS',
  'account/SET_AVATAR_FAILURE'
)<SharedImage, void, any>()

export const setRecoveryPhrase = createAction(
  'account/SET_RECOVERY_PHRASE',
  resolve => {
    return (recoveryPhrase: string) => resolve({ recoveryPhrase })
  }
)
