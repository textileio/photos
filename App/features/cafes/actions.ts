import { createAsyncAction } from 'typesafe-actions'

// Registering with a new cafe requires the cafe URL and string
// With the new SDK, the URL needs to be changed to the peer id
export const registerCafe = createAsyncAction(
  'cafes/REGISTER_CAFE_REQUEST',
  'cafes/REGISTER_CAFE_SUCCESS',
  'cafes/REGISTER_CAFE_ERROR'
)<
  {
    peerId: string
    token: string
    success?: () => void
  },
  string,
  {
    peerId: string
    error: any
  }
>()

// Deregistering from a cafe requires the cafe's peer id
export const deregisterCafe = createAsyncAction(
  'cafes/DEREGISTER_CAFE_REQUEST',
  'cafes/DEREGISTER_CAFE_SUCCESS',
  'cafes/DEREGISTER_CAFE_ERROR'
)<
  {
    id: string
    success?: () => void
  },
  string,
  {
    id: string
    error: any
  }
>()
