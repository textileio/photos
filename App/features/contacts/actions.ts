import { createAction } from 'typesafe-actions'
import { pb } from '@textile/react-native-sdk'
import Contacts from 'react-native-contacts'

export const getContactsRequest = createAction('GET_CONTACTS_REQUEST')

export const getContactsSuccess = createAction('GET_CONTACT_SUCCESS', (resolve) => {
  return (contacts: ReadonlyArray<pb.IContact>) => resolve({ contacts })
})

export const searchRequest = createAction('@contacts/SEARCH_REQUEST', (resolve) => {
  return (searchString: string) => resolve({ searchString })
})

export const searchStarted = createAction('@contacts/SEARCH_STARTED')

export const searchResultTextile = createAction('@contacts/SEARCH_RESULT_TEXTILE', (resolve) => {
  return (result: pb.IContact) => resolve({ result })
})

export const textileSearchComplete = createAction('@contacts/TEXTILE_SEARCH_COMPLETE')

export const searchResultsAddressBook = createAction('@contacts/SEARCH_RESULTS_ADDRESS_BOOK', (resolve) => {
  return (results: ReadonlyArray<Contacts.Contact>) => resolve({ results })
})

export const searchErrorTextile = createAction('@contacts/SEARCH_ERROR_TEXTILE', (resolve) => {
  return (error: any) => resolve({ error })
})

export const searchErrorAddressBook = createAction('@contacts/SEARCH_ERROR_ADDRESS_BOOK', (resolve) => {
  return (error: any) => resolve({ error })
})

export const clearSearch = createAction('@contacts/CLEAR_SEARCH')

export const addContactRequest = createAction('@contacts/ADD_CONTACT_REQUEST', (resolve) => {
  return (contact: pb.IContact) => resolve({ contact })
})

export const addContactSuccess = createAction('@contacts/ADD_CONTACT_SUCCESS', (resolve) => {
  return (contact: pb.IContact) => resolve({ contact })
})

export const addContactError = createAction('@contacts/ADD_CONTACT_ERROR', (resolve) => {
  return (contact: pb.IContact, error: any) => resolve({ contact, error })
})

export const clearAddContact = createAction('@contacts/CLEAR_ADD_CONTACT_ERROR', (resolve) => {
  return (contact: pb.IContact) => resolve({ contact })
})

export const authorInviteRequest = createAction('@contacts/AUTHOR_INVITE_REQUEST', (resolve) => {
  return (contact: Contacts.Contact) => resolve({ contact })
})
