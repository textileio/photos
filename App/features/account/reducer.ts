import { combineReducers } from 'redux'
import { ActionType, getType } from 'typesafe-actions'
import { IContact, ICafeSession } from '@textile/react-native-sdk'

import { SharedImage } from '../group/add-photo/models'
import * as actions from './actions'

export interface AccountState {
  readonly initialized: boolean, // splitting 'Preferences.onboarded' to within sdk 'initialized' and app specific 'onboarded'
  readonly chosenProfilePhoto: {
    readonly image?: SharedImage
    readonly data?: string
    readonly error?: string
  }
  readonly profile: {
    readonly value?: IContact
    readonly processing: boolean
    readonly error?: string
  }
  readonly peerId: {
    readonly value?: string
    readonly error?: string
  },
  readonly address: {
    readonly value?: string
    readonly error?: string
  }
  readonly avatar: {
    readonly error?: string
  }
  readonly recoveryPhrase: string,
  readonly cafeSessions: {
    readonly sessions: ReadonlyArray<ICafeSession>
    readonly processing: boolean
    readonly error?: string
  }
}

export type AccountAction = ActionType<typeof actions>

export default combineReducers<AccountState, AccountAction>({
  initialized: (state = false, action) => {
    return state
  },
  chosenProfilePhoto: (state = {}, action) => {
    switch (action.type) {
      case getType(actions.chooseProfilePhoto.success):
      case getType(actions.chooseProfilePhoto.failure):
        return { ...state, ...action.payload }
      case getType(actions.setAvatar.success):
      case getType(actions.cancelProfilePhotoUpdate):
        return {}
      default:
        return state
    }
  },
  profile: (state = { processing: true }, action) => {
    switch (action.type) {
      case getType(actions.refreshProfileRequest):
        return { ...state, processing: true }
      case getType(actions.refreshProfileSuccess):
        const { profile } = action.payload
        return { value: profile, processing: false }
      case getType(actions.profileError): {
        const obj = action.payload.error
        const error = obj.message as string || obj as string || 'unknown error'
        return { ...state, processing: false, error }
      }
      default:
        return state
    }
  },
  peerId: (state = {}, action) => {
    switch (action.type) {
      case getType(actions.refreshPeerIdSuccess):
        const { peerId } = action.payload
        return { value: peerId }
      case getType(actions.refreshPeerIdError): {
        const obj = action.payload.error
        const error = obj.message as string || obj as string || 'unknown error'
        return { ...state, error }
      }
      default:
        return state
    }
  },
  address: (state = {}, action) => {
    switch (action.type) {
      case getType(actions.refreshAddressSuccess):
        const { address } = action.payload
        return { value: address }
      case getType(actions.refreshAddressError): {
        const obj = action.payload.error
        const error = obj.message as string || obj as string || 'unknown error'
        return { ...state, error }
      }
      default:
        return state
    }
  },
  avatar: (state = {}, action) => {
    switch (action.type) {
      case getType(actions.setAvatar.failure): {
        const obj = action.payload
        const error = obj.message as string || obj as string || 'unknown error'
        return { ...state, error }
      }
      default:
        return state
    }
  },
  recoveryPhrase: (state = '', action) => {
    switch (action.type) {
      case getType(actions.setRecoveryPhrase):
        return action.payload.recoveryPhrase
      default:
        return state
    }
  },
  cafeSessions: (state = { processing: false, sessions: [] }, action) => {
    switch (action.type) {
      case getType(actions.refreshCafeSessionsRequest):
        return { ...state, processing: true, error: undefined }
      case getType(actions.cafeSessionsSuccess):
        return { sessions: action.payload.sessions , processing: false, error: undefined }
      case getType(actions.cafeSessionsError):
        const obj = action.payload.error
        const error = obj.message as string || obj as string || 'unknown error'
        return { ...state, processing: false, error }
      default:
        return state
    }
  }
})
