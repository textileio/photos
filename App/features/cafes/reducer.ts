import { combineReducers } from 'redux'
import { PersistConfig, persistReducer } from 'redux-persist'
import { AsyncStorage } from 'react-native'
import { ActionType, getType } from 'typesafe-actions'

import * as actions from './actions'
import { CafeSessionsData, Cafes, Cafe } from './models'

export interface CafesState {
  readonly cafes: Cafes
  readonly cafeSessions: {
    readonly sessions: CafeSessionsData
    readonly processing: boolean
    readonly error?: string
  }
}

export type CafesAction = ActionType<typeof actions>

const persistConfig: PersistConfig = {
  key: 'cafes',
  storage: AsyncStorage,
  whitelist: ['cafes'],
  debug: false
}

const reducer = combineReducers<CafesState, CafesAction>({
  cafes: (state = {}, action) => {
    switch (action.type) {
      // Update our list of registered cafes with cafes that are actuall registered to
      // catch cafe data for cafes that were registered before we added this redux state
      case getType(actions.getCafeSessions.success): {
        const { sessions } = action.payload
        const cafes = sessions.reduce((cafes, session) => {
          const { peer } = session.cafe
          const updated: Cafe = {
            ...cafes[peer],
            peerId: peer,
            state: 'registered'
          }
          return {
            ...cafes,
            [peer]: updated
          }
        }, state)
        return cafes
      }
      case getType(actions.registerCafe.request): {
        const { peerId, token } = action.payload
        return {
          ...state,
          [peerId]: { peerId, token, state: 'registering' }
        }
      }
      case getType(actions.registerCafe.success): {
        const peerId = action.payload
        return { ...state, [peerId]: { ...state[peerId], state: 'registered' } }
      }
      case getType(actions.registerCafe.failure): {
        const { peerId, error } = action.payload
        const errorMessage =
          (error.message as string) || (error as string) || 'unknown error'
        return {
          ...state,
          [peerId]: {
            ...state[peerId],
            error: errorMessage
          }
        }
      }
      case getType(actions.deregisterCafe.request): {
        const { peerId } = action.payload
        return {
          ...state,
          [peerId]: { ...state[peerId], state: 'deregistering' }
        }
      }
      case getType(actions.deregisterCafe.success): {
        const { [action.payload]: deregistered, ...newState } = state
        return newState
      }
      case getType(actions.deregisterCafe.failure): {
        const { peerId, error } = action.payload
        const errorMessage =
          (error.message as string) || (error as string) || 'unknown'
        return {
          ...state,
          [peerId]: {
            ...state[peerId],
            error: errorMessage
          }
        }
      }
      default:
        return state
    }
  },
  cafeSessions: (state = { sessions: {}, processing: false }, action) => {
    switch (action.type) {
      case getType(actions.getCafeSessions.request): {
        return { ...state, processing: true, error: undefined }
      }
      case getType(actions.getCafeSessions.success): {
        const { sessions } = action.payload
        const cafeSessionsData = sessions.reduce(
          (cafeSessionsData, session) => {
            return {
              ...cafeSessionsData,
              [session.cafe.peer]: { session, processing: false }
            }
          },
          {} as CafeSessionsData
        )
        return { sessions: cafeSessionsData, processing: false }
      }
      case getType(actions.getCafeSessions.failure): {
        const { error } = action.payload
        const message =
          (error.message as string) || (error as string) || 'unknown error'
        return { ...state, error: message, processing: false }
      }
      case getType(actions.refreshCafeSession.request): {
        const { peerId } = action.payload
        return {
          ...state,
          sessions: {
            ...state.sessions,
            [peerId]: { ...state.sessions[peerId], processing: true }
          }
        }
      }
      case getType(actions.refreshCafeSession.success): {
        const { session } = action.payload
        const peerId = session.cafe.peer
        return {
          ...state,
          sessions: {
            ...state.sessions,
            [peerId]: { session, processing: false }
          }
        }
      }
      case getType(actions.refreshCafeSession.failure): {
        const { peerId, error } = action.payload
        const message =
          (error.message as string) || (error as string) || 'unknown error'
        return {
          ...state,
          sessions: {
            ...state.sessions,
            [peerId]: {
              ...state.sessions[peerId],
              processing: false,
              error: message
            }
          }
        }
      }
      default:
        return state
    }
  }
})

export default persistReducer(persistConfig, reducer)
