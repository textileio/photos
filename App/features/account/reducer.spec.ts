import { IContact } from '@textile/react-native-sdk'

import * as actions from './actions'
import reducer from './reducer'
import { SharedImage } from '../group/add-photo/models'

const initialState = reducer(undefined, {} as any)
const profile: IContact = {
  address: 'address',
  name: 'username',
  avatar: 'avatar',
  peers: [],
  threads: []
}
const peerId = 'peerId'
const error = 'error'
const recoveryPhrase = 'recoveryPhrase'
const accountSeed = 'accountSeed'

describe('account', () => {
  describe('initial state', () => {
    it('should match snapshot', () => {
      expect(initialState).toMatchSnapshot()
    })
  })
  describe('profile', () => {
    it('should be processing from refresh', () => {
      const state0 = reducer(initialState, actions.refreshProfileRequest())
      expect(state0.profile.processing).toBeTruthy()
    })
    it('should be processing from setUsername', () => {
      const state0 = reducer(
        initialState,
        actions.setUsernameRequest('username')
      )
      expect(state0.profile.processing).toBeTruthy()
    })
    it('should update profile', () => {
      const state0 = reducer(
        initialState,
        actions.refreshProfileSuccess(profile)
      )
      expect(state0.profile.value).toEqual(profile)
      expect(state0.profile.processing).toBeFalsy()
    })
    it('should track profile error', () => {
      const state0 = reducer(initialState, actions.profileError(error))
      expect(state0.profile.error).toEqual(error)
      expect(state0.profile.processing).toBeFalsy()
    })
  })
  describe('peer id', () => {
    it('should not change state for request', () => {
      const state0 = reducer(initialState, actions.refreshPeerIdRequest())
      expect(state0).toEqual(initialState)
    })
    it('should update peer id', () => {
      const state0 = reducer(initialState, actions.refreshPeerIdSuccess(peerId))
      expect(state0.peerId.value).toEqual(peerId)
    })
    it('should track peer id error', () => {
      const state0 = reducer(initialState, actions.refreshPeerIdError(error))
      expect(state0.peerId.error).toEqual(error)
    })
  })
  describe('account seed', () => {
    it('should update account seed', () => {
      const state0 = reducer(
        initialState,
        actions.refreshAccountSeed.success(accountSeed)
      )
      expect(state0.accountSeed.value).toEqual(accountSeed)
    })
  })
  describe('avatar', () => {
    it('should not change state for request', () => {
      const state0 = reducer(
        initialState,
        actions.setAvatar.request({} as SharedImage)
      )
      expect(state0).toEqual(initialState)
    })
    it('should track avatar error', () => {
      const state0 = reducer(initialState, actions.setAvatar.failure(error))
      expect(state0.avatar.error).toEqual(error)
    })
  })
  describe('recovery phrase', () => {
    it('should track revocery phrase', () => {
      const state0 = reducer(
        initialState,
        actions.setRecoveryPhrase(recoveryPhrase)
      )
      expect(state0.recoveryPhrase).toEqual(recoveryPhrase)
    })
  })
})
