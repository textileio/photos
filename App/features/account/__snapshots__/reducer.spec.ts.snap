// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`account initial state should match snapshot 1`] = `
Object {
  "address": Object {},
  "avatar": Object {},
  "cafeSessions": Object {
    "processing": false,
    "sessions": Array [],
  },
  "chosenProfilePhoto": Object {},
  "initialized": false,
  "peerId": Object {},
  "profile": Object {
    "processing": true,
  },
  "recoveryPhrase": "",
}
`;
